File: prime_numbers.py
# compute primes up to max number using the Sieve algorithm
# https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
#

import sys
import math
import numpy as np
# from numpy import ones, dtype, np.bool
from datetime import *


def print_usage(name):
    print('python3 {} <max_number>'.format(name))

def create_boolean_list(max: int):
    numbers = np.ones(int(max/2), dtype=np.bool)
    numbers[1] = False
    return numbers

def mark_multiples_of_prime(numbers, prime_index: int, max: int):
    prime = (prime_index*2)-1
    array_length = len(numbers)
    
    # print('Checking prime:{}'.format(prime))

    # the for loop will start at one multiple past the current prime
    # it will walk the list of numbers by multiples of prime
    # this logic removes the need for any if statements inside the loop
    for i in range(prime_index+prime, array_length, prime):
        numbers[i] = False

def compute_primes(max: int):
    start_time = datetime.now()

    numbers = create_boolean_list(max)

    # print('Numbers = {}'.format(numbers))

    print('Computing primes from 0 to {}'.format(max))

    half_length = int(len(numbers)/2)

    for index in range(int(1), half_length):
        if (numbers[index] ==  True):
            # print('index: {}, prime: {}'.format(index, (index*2)-1))
            mark_multiples_of_prime(numbers, index, max)

    primes = [i for i in numbers if i == True]

    mid_time = datetime.now()

    # print('Primes = {}'.format(primes))

    finish_time = datetime.now()

    print('Start: {}, Mid: {}, Finish: {}'.format(start_time, mid_time, finish_time))
    print('Compute time: {}. \nFound {} prime numbers less than {}'.format(mid_time - start_time, len(primes), max))
    # print('Print time: {}'.format(finish_time - mid_time))

def main():
    if (len(sys.argv) > 1):
        compute_primes(int(sys.argv[1]))
    else:
        print_usage(sys.argv[0])


if __name__ == "__main__":
    main()

File: prime_numbers.jl
# compute primes up to max using the Sieve algorithm
# https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes
# https://docs.julialang.org/en/v1/


import Dates

function print_usage()
  println(string("julia ", PROGRAM_FILE, " max_prime"))  
end

function make_array(max)
  return trues(Integer(max/2))
end

function remove_multiples(numbers, prime_index, max)

  prime::Int64 = Int64((prime_index*2)-1)
  array_length::Int64 = length(numbers)

  # println("called remove_multiple(", prime_index, ") = ", prime, ", length: ", array_length)

  # the for loop will start at one multiple past the current prime
  # it will walk the list of numbers by multiples of prime
  # this logic removes the need for any if statements inside the loop
  @inbounds for current_index::Int64 in prime_index+prime : prime : array_length
    numbers[current_index] = 0
  end
end

function compute_primes(max::Int64)
  start_time = Dates.now()

  numbers = make_array(max)

  # println(numbers)
  # println("Length: ", length(numbers))

  println("Computing primes from 0 to ", max)

  half_length::Int64 = Int64(max/4)

  @inbounds for i = 2:half_length
    if numbers[i] > 0
      remove_multiples(numbers, i, max)
    end
  end


  primes = filter(x -> x>0, numbers)

  mid_time = Dates.now()

  # println("Primes: ", primes)

  finish_time = Dates.now()

  println("Start: ", start_time, ", Mid: ", mid_time, ", Finish: ", finish_time)
  println("Compute time: ", (mid_time - start_time))
  println("Found: ",  length(primes), " prime numbers under: ", max)
end

function main()
  if length(ARGS) < 1
    print_usage()
  else
    compute_primes(parse(Int64, ARGS[1]))
  end

end

main()
